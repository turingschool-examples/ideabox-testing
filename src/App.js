import React, { Component } from 'react';
import Ideas from './Ideas';
import Form from './Form';
import './App.css';

class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      error: ''
    }
  }

  componentDidMount() {
    fetch('http://localhost:3001/api/v1/ideas')
      .then(response => response.json())
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  addIdea = (newIdea) => {
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ ...newIdea })
    };

    fetch('http://localhost:3001/api/v1/ideas', options)
      .then(response => response.json())
      .then(response => fetch(`http://localhost:3001/api/v1/ideas/${response.id}`))
      .then(response => response.json())
      .then(newIdea => this.setState({ ideas: [...this.state.ideas, newIdea] }))
      .catch(error => this.setState({ error: error.message }))
  }

  deleteIdea = (id) => {
    const options = {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json'
      }
    }

    fetch(`http://localhost:3001/api/v1/ideas/${id}`, options)
      .then(() => fetch('http://localhost:3001/api/v1/ideas'))
      .then(response => response.json())
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  render() {
    return(
      <main className='App'>
        <h1>IdeaBox</h1>
        <Form addIdea={this.addIdea} />
        {this.state.error && <h2>{this.state.error}</h2>}
        <Ideas ideas={this.state.ideas} deleteIdea={this.deleteIdea} />
      </main>
    )
  }
}

export default App;
