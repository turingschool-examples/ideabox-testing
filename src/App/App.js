import React, { Component } from 'react';
import Form from '../Form/Form';
import Ideas from '../Ideas/Ideas'
import { getIdeas, postIdea, deleteIdea } from '../apiCalls';
import './App.css';

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      ideas: [],
      isLoading: true,
      error: ''
    };
  }

  componentDidMount() {
    getIdeas()
      .then(ideas => this.setState({ ideas, isLoading: false }))
      .catch(error => this.setState({
        isLoading: false,
        error: error.message
      })
    );
  }

  addIdea = newIdea => {
    postIdea(newIdea)
      .then(idea => this.setState({
        ideas: [...this.state.ideas, idea]
      }))
      .catch(error => this.setState({ error: error.message }))
  }

  removeIdea = id => {
    deleteIdea(id)
      .then(ideas => this.setState({ ideas }))
      .catch(error => this.setState({ error: error.message }));
  }

  render() {
    const { ideas, isLoading, error } = this.state;
    return (
      <main className="App">
        <h1>IdeaBox</h1>
        <Form addIdea={this.addIdea} />
        {isLoading && <img
          src={'https://www.gearbubble.com/assets/loader_large.gif'}
          alt={''}
        />
        }
        {error && <h2>{error}</h2>}
        <Ideas
          ideas={ideas}
          removeIdea={this.removeIdea}
        />
      </main>
    )
  }
}