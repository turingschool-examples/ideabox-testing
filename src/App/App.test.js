import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should update state with an idea when addIdea is called', () => {
    // Setup
    const wrapper = shallow(<App />);
    const mockIdea = { title: 'sweaters for pugs', description: 'why not?', id: Date.now() };
    const expected = [mockIdea];

    // Expectation
    expect(wrapper.state('ideas')).toEqual([]);

    // Execution
    wrapper.instance().addIdea(mockIdea);

    // Expectation
    expect(wrapper.state('ideas')).toEqual(expected);
  });

  it('should remove an idea from state when removeIdea is called', () => {
    const wrapper = shallow(<App />);
    const mockIdeas = [
      {id: 1, title: 'Sweaters for pugs', description: 'To keep them warm'},
      {id: 2, title: 'A romcom', description: 'But make it ghosts'},
      {id: 3, title: 'A game show called Ether/Or', description: 'When you lose you get chloroformed'},
    ];
    const expected = [
      {id: 1, title: 'Sweaters for pugs', description: 'To keep them warm'},
      {id: 3, title: 'A game show called Ether/Or', description: 'When you lose you get chloroformed'},
    ]

    wrapper.instance().setState({ ideas: mockIdeas });
    wrapper.instance().removeIdea(2);

    expect(wrapper.state('ideas')).toEqual(expected);
  });
});
