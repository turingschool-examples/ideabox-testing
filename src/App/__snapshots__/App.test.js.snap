// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<main
  className="App"
>
  <h1>
    IdeaBox
  </h1>
  <Form
    addIdea={[Function]}
  />
  <img
    alt=""
    src="https://www.gearbubble.com/assets/loader_large.gif"
  />
  <Ideas
    ideas={Array []}
    removeIdea={[Function]}
  />
</main>
`;
